File: level2/main.js
const object = require('/Users/dustinstender/Downloads/serpentini - candidat/level2/data/input.json');

// calculates commission given the total amount and their objective
function computeCommission(amount, objective) {
	let total = 0;
	if (amount > objective) {
		total += (amount - objective) * 0.15;
		// this is for both the first and second tier
		total += (objective / 2) * 0.15;
	} else if (amount > objective / 2) {
		total += (amount - objective / 2) * 0.1;
		total += (objective / 2) * 0.05;
	} else {
		total += amount * 0.05;
	}
	return total;
}

// calculates total amount generated for each salesperson
function totalAmount(jsonFile) {
	const totalUsers = jsonFile.users.length;
	let totalAmountArray = Array(totalUsers).fill(0);
	jsonFile.deals.forEach((element) => {
		totalAmountArray[element.user - 1] += element.amount;
	});
	return totalAmountArray;
}

// creates an array of commission
function createCommissionArray(jsonFile) {
	let totalAmountArray = totalAmount(jsonFile);
	let commissionAmountArray = Array(totalAmountArray.length).fill(0);
	// console.log(jsonFile.users[0].objective);
	for (let i = 0; i < totalAmountArray.length; ++i) {
		commissionAmountArray[i] += computeCommission(
			totalAmountArray[i],
			jsonFile.users[i].objective
		);
	}
	return commissionAmountArray;
}

// converts input to the final output object
function convertToObject(input) {
	const commissionArray = createCommissionArray(input);
	const object = commissionArray.map((person, index) => ({
		user_id: input.users[index].id,
		commission: person,
	}));
	return {
		commissions: object,
	};
}
console.log(convertToObject(object));

File: level2/main.py
object = {
	'users': [
		{ 'id': 1, 'name': 'Nicolas', 'objective': 1000 },
		{ 'id': 2, 'name': 'Math', 'objective': 500 },
	],
	'deals': [
		{ 'id': 1, 'amount': 500, 'user': 1 },
		{ 'id': 2, 'amount': 1000, 'user': 2 },
		{ 'id': 3, 'amount': 800, 'user': 1 },
	],
}

# calculates commission given the total amount and their objective
def caluclate_commission(amount, objective):
    total = 0
    if amount > objective:
        total += (amount - objective) * 0.15
        total += (objective / 2) * 0.15
    elif amount > (objective / 2):
        total += (amount - objective / 2) * 0.1
        total += (objective / 2) * 0.05
    else:
        total += amount * 0.05
    return total

# calculates total amount generated for each salesperson
def total_amount(object):
    total_users = len(object['users'])
    total_amount_list = [0] * total_users
    for deal in object['deals']:
        total_amount_list[deal['user'] - 1] += deal['amount']
    return total_amount_list

# creates an array of commission
def create_commissions_list(object):
    total_amount_list = total_amount(object)
    commission_amount_list = [0] * len(total_amount_list)
    for i in range(len(total_amount_list)):
        commission_amount_list[i] += caluclate_commission(total_amount_list[i], object['users'][i]['objective'])
    return commission_amount_list

# converts input to the final output dictionary
def convert_to_dictionary(object):
    commission_list = create_commissions_list(object)
    dict = [{'user_id':i + 1 ,'commission': x} for i,x in enumerate(commission_list)]
    return {
        "commissions": dict
    }

print(convert_to_dictionary(object))
